### ISSUE TITLE
Fix trailing commas in sequences

### ISSUE THREAD

**Issue by jrappen** [jrappen/sublime-json#1]
We want to do this without using branching, as that maxes out at 128 lines. That is unacceptable for JSON.

**Comment by deathaxe** [comment]
You want to highlight trailing commas illegal?

Why is the line limit a problem with regards to branching?

Just check whether the following char is `]` or `}`. If not highlight illegal. Branching would just be needed to find closing brackets on the next line(s). The only thing which might come into the way are huge comments. But that's acceptable.

**Comment by jrappen** [comment]
Yes.

Because `(any value) ((,) (any value))*`.

So if we for example assume having an array of arrays, be it computer generated or not, we could easily end up hitting that 128 lines limit.

**Comment by deathaxe** [comment]
What I had in mind is just the following. Haven't benchmarked it against your current solution though.

```yaml
  sequences:
    - include: empty-sequences
    - match: \[
      scope: punctuation.definition.sequence.begin.json
      push:
        - sequence-body
        - illegal-separator

  sequence-body:
    - meta_scope: meta.sequence.list.json
    - match: \]
      scope: punctuation.definition.sequence.end.json
      pop: 1
    - include: any
    - include: sequence-separators
    - match: '[^\s\]]'
      scope: invalid.illegal.expected-separator.json

  sequence-separators:
    - match: (?=,)
      branch_point: sequence-separators
      branch:
        - sequence-separator
        - illegal-separator

  sequence-separator:
    - match: ','
      scope: punctuation.separator.sequence.json
      set: expect-value

  expect-value:
    - match: (?=\])
      fail: sequence-separators
    - include: illegal-separator

  illegal-separator:
    - match: ','
      scope: invalid.illegal.unexpected-separator.json
    - match: (?=\S)
      pop: 1

  empty-sequences:
    - match: (\[)\s*(\])
      scope: meta.sequence.list.empty.json
      captures:
        1: punctuation.definition.sequence.begin.json
        2: punctuation.definition.sequence.end.json
```

**Comment by jrappen** [comment]
I haven't benchmarked anything yet either. I'm still collecting possible solutions at this point.

I'll look into this a bit later this afternoon.

**Comment by jrappen** [comment]
fixed in #4, compare 7b559e6
